cocotb/types/logic.py

```diff
@@ -0,0 +1,279 @@
+# Copyright cocotb contributors
+# Licensed under the Revised BSD License, see LICENSE for details.
+# SPDX-License-Identifier: BSD-3-Clause
+from typing import Any, Optional
+from functools import lru_cache
+
+
+class Logic:
+    r"""
+    Model of a 4-value (``0``, ``1``, ``X``, ``Z``) datatype commonly seen in HDLs.
+
+    .. currentmodule:: cocotb.types
+
+    This is modeled after (System)Verilog's 4-value ``logic`` type.
+    VHDL's 9-value ``std_ulogic`` type maps to this type by treating weak values as full strength values
+    and treating "uninitialized" (``U``) and "don't care" (``-``) as "unknown" (``X``).
+
+    :class:`Logic` can be converted to and from :class:`int`, :class:`str`, :class:`bool`, and :class:`Bit`
+    by using the appropriate constructor syntax.
+    The list of values convertable to :class:`Logic` includes
+    ``0``, ``1``, ``True``, ``False``, ``"0"``, ``"1"``, ``"X"``, ``"Z"``, ``Bit('0')``, and ``Bit('1')``.
+    For a comprehensive list of values that can be converted into :class:`Logic` see :file:`tests/pytest/test_logic.py`.
+
+    .. code-block:: python3
+
+        >>> Logic("X")
+        Logic('X')
+        >>> Logic(True)
+        Logic('1')
+        >>> Logic(1)
+        Logic('1')
+        >>> Logic(Bit(0))
+        Logic('0')
```

Comment:
What's your thought on the canonical spelling (and `__repr__`) being `Logic.X`, Logic.Z`, `Logic._0`, and `Logic._1`?

Reply:
I'm guessing this would also tie into a singleton cache and keeping the values on the type like enum does? I had that in a previous implementation. That's fine by me.

Reply:
> I'm guessing this would also tie into a singleton cache and keeping the values on the type like enum does?

Yes, exactly. I'm not particularly attached to `Logic._0` over `Logic(0)`, but I think `Logic.X` is a clear readability improvement over `Logic("X")`, and should be faster too.

Reply:
I think there is an issue with where `Bit.X` would work and would return `Logic("X")`.

Reply:
Oh, good point. That's not too awful to fix:
```python
In [112]: class StaticOnlyProp:
     ...:     def __init__(self, fget):
     ...:         self.fget = fget
     ...:     def __set_name__(self, cls, name):
     ...:         self.__cls = cls
     ...:     def __get__(self, instance, cls):
     ...:         if cls is not self.__cls or instance is not None:
     ...:             raise AttributeError
     ...:         return self.fget()
     ...:
     ...:

In [113]: class Logic:
     ...:     @StaticOnlyProp
     ...:     def X():
     ...:         return "X"
     ...:

In [114]: class Bit(Logic): pass

In [115]: Logic.X
Out[115]: 'X'

In [116]: Logic().X  # this is nonsense
AttributeError

In [117]: Bit.X  # this is nonsense
AttributeError
```

Perhaps that's enough of an argument to leave it for a follow-up; your current spelling is likely something we want to have anyway, and changing a preferred spelling is a much easier thing to do later than removing a spelling.